src/instrumentation.ts

```diff
@@ -48,3 +67,66 @@ export function initTelemetry(defaultServiceName: string = "hasura-ndc", default
 export function isInitialized(): boolean {
   return sdk !== null;
 }
+
+export const USER_VISIBLE_SPAN_ATTRIBUTE: Attributes = {
+  "internal.visibility": "user",
+};
+
+export function withActiveSpan<TReturn>(
+  tracer: Tracer,
+  name: string,
+  func: (span: Span) => TReturn,
+  attributes?: Attributes
+): TReturn {
+  return withInternalActiveSpan(tracer, name, func, attributes ? { ...USER_VISIBLE_SPAN_ATTRIBUTE, ...attributes } : USER_VISIBLE_SPAN_ATTRIBUTE);
+}
+
+export function withInternalActiveSpan<TReturn>(
+  tracer: Tracer,
+  name: string,
+  func: (span: Span) => TReturn,
+  attributes?: Attributes
+): TReturn {
+  return tracer.startActiveSpan(name, (span) => {
+    if (attributes) span.setAttributes(attributes);
+
+    const handleError = (err: unknown) => {
+      if (err instanceof Error || typeof err === "string") {
+        span.recordException(err);
+      }
+      span.setStatus({ code: SpanStatusCode.ERROR });
+      span.end();
+    };
+
+    try {
+      const retval = func(span);
+      // If the function returns a Promise, then wire up the span completion to
+      // the completion of the promise
+      if (
+        typeof retval === "object" &&
+        retval !== null &&
+        "then" in retval &&
+        typeof retval.then === "function"
```

Comment:
It seems like this logic isn't really applicable to "getCapabilities" etc.

Reply:
Could this be simplified? Or just have a promise and non-promise version since you should know what kind of function it is wrapping?

Reply:
This code is basically stolen from a prototype version of something they're thinking of adding to OpenTelemetry itself. I'd rather not change it.